### PR TITLE
New rule deep-exit

### PR DESCRIPTION
**Pull request by chavacava**
`deep-exit` looks for program exits in functions other than `main()` or `init()`

Opinionated guidelines suggest that programs should exit only from `main()` or `init()` functions.

### CODE REVIEW

File: fixtures/deep-exit.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    log.Fatalln("v ...interface{}")
+}
+
+func bar() {
+    log.Fatal(1) // MATCH /calls to log.Fatal function should be made only in main() or init() functions/
+}
+
+func bar2() {
+    syscall.Exit(1) // MATCH /calls to syscall.Exit function should be made only in main() or init() functions/
+}
```
Comment (mgechev) [inline fixtures/deep-exit.go]: newline

File: rule/deep-exit.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	if stmt, ok := node.(*ast.FuncDecl); ok {
+		w.updateIgnore(stmt)
+		return w
+	}
+
+	if w.ignore {
+		return w
+	}
+
+	if se, ok := node.(*ast.ExprStmt); ok {
```
Comment (mgechev) [inline rule/deep-exit.go]: It'll be easier to follow if it's:

```go
```
if se, ok := node.(*ast.ExprStmt); !ok {
  return nil
}
if ce, ok := se.X.(*ast.CallExpr); !ok {
  return nil
}
...

File: rule/deep-exit.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+		if ce, ok := se.X.(*ast.CallExpr); ok { // it's a function call
+			if fc, ok := ce.Fun.(*ast.SelectorExpr); ok {
+				if id, ok := fc.X.(*ast.Ident); ok {
+					fn := fc.Sel.Name
+					pkg := id.Name
+					if w.exitFunctions[pkg][fn] { // it's a call to an exit function
```
Comment (mgechev) [inline rule/deep-exit.go]: `w.exitFunctions[pkg]` could be `nil`

Comment (chavacava) [reply rule/deep-exit.go]: :+1:

File: rule/deep-exit.go (hunk 2 above)
Comment (chavacava) [reply rule/deep-exit.go]: :+1:

File: fixtures/deep-exit.go (hunk 1 above)
Comment (chavacava) [reply fixtures/deep-exit.go]: :+1:

File: rule/deep-exit.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+	fn := fc.Sel.Name
+	pkg := id.Name
+	if w.exitFunctions[pkg] != nil && w.exitFunctions[pkg][fn] { // it's a call to an exit function
+		w.onFailure(lint.Failure{
+			Confidence: 1,
+			Node:       ce,
+			Category:   "bad practice",
+			URL:        "#deep-exit",
+			Failure:    fmt.Sprintf("calls to %s.%s function should be made only in main() or init() functions", pkg, fn),
```
Comment (mgechev) [inline rule/deep-exit.go]: Can we try to remove the passive:

`call %s.%s only in main() and init() functions`

**Pull request merged**